### PR TITLE
Drop closure environment

### LINKED ISSUE #138: Destructor for Rust closure environment doesn't run
This is the source code of the test:

https://github.com/mkpankov/hlua-test/blob/master/src/main.rs

Why dropping `lua` doesn't `drop` `f`?

Is this a known limitation?

Is there a workaround?

### PR DESCRIPTION
**Pull request by mkpankov**
This fixes #138 . However, see comments for newly added test.

### CODE REVIEW

**Comment by tomaka** [comment]
Your test always passes because the closure doesn't contain `move`.

I'd even suggest using an `Arc<Foo>` instead of a `Foo` so that you are sure that it's not the local variable that is destroyed.

**Comment by jonas-schievink** [comment]
You can also have the destructor increment a static variable and check its expected value in multiple places

**Comment by mkpankov** [comment]
Don't merge this, still WIP.

Using `Arc` or `Rc` starts to cause SIGSEGV, investigating.

**Comment by mkpankov** [comment]
Ok, so it segfaults because for closure userdata, I don't push a `TypeId` and then `destructor_wrapper` panics here:
https://github.com/tomaka/hlua/blob/master/hlua/src/userdata.rs#L24

Gonna add another destructor wrapper for closures specifically, as I don't want to require `'static` on `Z` in `Push for Function<Z,...`. Closure pushed to Lua must just outlive Lua, not be static.

**Comment by mkpankov** [comment]
Should be good to go now.

File: hlua/src/functions_write.rs [hunk 1]
```diff
@@ -496,4 +548,28 @@ mod tests {
...
+        struct Foo { };
+        impl Drop for Foo {
+            fn drop(&mut self) {
+                unsafe {
+                    DID_DESTRUCTOR_RUN = true;
+                }
+            }
+        }
+        {
+            let foo = ::std::sync::Arc::new(Foo { });
```
Comment (tomaka) [inline hlua/src/functions_write.rs]: Please use `use std::sync::Arc;` at the root of the tests module instead.

**Comment by tomaka** [comment]
Thanks, looks good other than the nit.

File: hlua/src/functions_write.rs (hunk 1 above)
Comment (mkpankov) [inline hlua/src/functions_write.rs]: Done

**Pull request merged**